drivers/io/io_block.c

```diff
@@ -0,0 +1,298 @@
+	.dev_init	= block_dev_init,
+	.dev_close	= block_dev_close,
+};
+
+static io_dev_info_t block_dev_info = {
+	.funcs	= &block_dev_funcs,
+	.info	= (uintptr_t)NULL,
+};
+
+io_type_t device_type_block(void)
+{
+	return IO_TYPE_BLOCK;
+}
+
+static int block_open(io_dev_info_t *dev_info, const uintptr_t spec,
+		      io_entity_t *entity)
+{
+	io_block_entity_t *block;
+	io_block_spec_t *region;
+
+	assert((dev_info->info != (uintptr_t)NULL) &&
+	       (spec != (uintptr_t)NULL) &&
+	       (entity->info == (uintptr_t)NULL));
+
+	block = (io_block_entity_t *)spec;
+	region = &block->region;
+	memcpy((void *)&block->buffer, (void *)dev_info->info,
+	       sizeof(io_block_spec_t));
+
+	assert((block->block_size > 0) &&
+	       ((block->block_size & MIN_BLOCK_MASK) == 0) &&
+	       ((region->offset % block->block_size) == 0) &&
+	       ((region->length % block->block_size) == 0) &&
+	       ((block->buffer.offset % block->block_size) == 0) &&
+	       ((block->buffer.length % block->block_size) == 0));
+
+	block->file_pos = 0;
+	block->lba = (block->file_pos + region->offset) / block->block_size;
+	entity->info = spec;
+	return 0;
+}
+
+/* parameter offset is relative address at here */
+static int block_seek(io_entity_t *entity, int mode, ssize_t offset)
+{
+	io_block_entity_t *block;
+	io_block_spec_t *region;
+
+	assert(entity->info != (uintptr_t)NULL);
+
+	block = (io_block_entity_t *)entity->info;
+	region = &block->region;
+
+	assert(offset < block->region.length);
+
+	block->file_pos += offset;
+	block->lba = (block->file_pos + region->offset) / block->block_size;
+	assert(block->lba >= 0);
+	return 0;
+}
+
+static int block_read(io_entity_t *entity, uintptr_t buffer, size_t length,
+		      size_t *length_read)
+{
+	io_block_entity_t *block;
+	io_block_spec_t *region, *buf;
+	size_t aligned_length, skip, count, left, padding;
+
+	assert(entity->info != (uintptr_t)NULL);
+	block = (io_block_entity_t *)entity->info;
+	region = &block->region;
+	assert((length <= region->length) &&
+	       (length > 0) &&
+	       (block->ops.read != 0));
+
+	buf = &block->buffer;
+	skip = block->file_pos % block->block_size;
+	aligned_length = ((skip + length) + (block->block_size - 1)) &
+			 ~(block->block_size - 1);
+	padding = aligned_length - (skip + length);
+	left = aligned_length;
+	do {
+		if (left >= buf->length) {
+			/* Since left is larger, it's impossible to padding. */
+			if (skip)
+				count = block->ops.read(block->lba, buf->offset,
+							buf->length);
+			else
+				count = block->ops.read(block->lba, buffer,
+							buf->length);
+			assert(count == buf->length);
+			block->file_pos += count - skip;
+			if (skip) {
+				memcpy((void *)buffer,
+				       (void *)(buf->offset + skip),
+				       count - skip);
+			}
+			left = left - (count - skip);
+		} else {
+			if (skip || padding)
+				count = block->ops.read(block->lba, buf->offset,
+							left);
+			else
+				count = block->ops.read(block->lba, buffer,
+							left);
```

Comment:
So there are 4 different ways that the underlying `ops.read()` function is called. It's very difficult to understand what all the different cases are for without more comments.